### DISCUSSION TITLE
Several ComboBox controls with one DataSource object affect each other on change

### DISCUSSION THREAD

**Discussion by shakir-timur** [dotnet/winforms#4627] [Q&A] [answered]
* .NET Core Version: 3.1

* Have you experienced this same bug with .NET Framework?:  Yes

**Problem description:**
Binding several ComboBoxes to one object causes somewhat unobvious side effects: when selecting value of one causes others to change theirs too and fire events like SelectedIndexChanged on every other ComboBoxes.
This can probably affect other elements too, not just ComboBoxes.

Such behavior doesn't seem to be documented and, imo, should be fixed.

Also some other people's encounters:
https://stackoverflow.com/questions/4344366/multiple-combo-boxes-with-the-same-data-source-c/
https://stackoverflow.com/questions/33108993/multiple-combobox-with-single-datasource/

**Expected behavior:**
All ComboBoxes act independently of each other.

**Minimal repro:**
Put two ComboBoxes and two Labels on a form. Change codebehind to something like this:

        public Form1()
        {
            InitializeComponent();

            var array = new List<string>() { "one", "two" };
            comboBox1.DataSource = array;
            comboBox2.DataSource = array;
        }

        int counter = 0;

        private void comboBox1_SelectedIndexChanged(object sender, EventArgs e)
        {
            counter++;
            label1.Text = counter.ToString();
        }

        private void comboBox2_SelectedIndexChanged(object sender, EventArgs e)
        {
            counter++;
            label2.Text = counter.ToString();
        }

**Comment by weltkante** [comment, marked as answer]
Both your linked discussions have the correct solutions, I don't believe there is a bug.

When you are assigning lists directly as data source WinForms will track the current selection for you, to support scenarios like master/detail views where the controls of the detail area need to respect the current selection of the master list. Obviously this can only work when the binding system coordinates between controls.

If you want separate selection logic you should use separate `BindingSource`s instead of directly binding against a list. WinForms is a strongly designer-oriented framework so the "expected" way to do this in the designer is to place a `BindingSource` in the designer for each "selection context" you need, prepare your bindings in the designer against the `BindingSource`, and then connect that to your data source, rather than connecting the data sources of individual controls directly to your controls.

Of course you can also do all of this in code, but here you also have an alternate solution (also suggested in your linked threads) to explicitly instruct WinForms to separate the controls by assigning them separate `BindingContext` instances. The binding context is what stores the implicitly associated data like the current selection, so if you have separate contexts they will also have separate selection.
